### PR TITLE
extract metadata (NcML XML) from NetCDF/HDF5 files, new "requirements" option for external tools

### LINKED ISSUE #9153: Extract metadata from NetCDF and HDF5 files as XML in NcML format
Assuming PR #9152 is merged we'll have a library in place to start extracting XML from NetCDF and HDF5 files.

The supported XML format is called NcML and is described here: https://docs.unidata.ucar.edu/netcdf-java/current/userguide/ncml_overview.html

Yesterday there was general agreement among devs that it would be fine to save the XML as a derivative or aux file.

This will open the door for previewing the file as raw XML to start.

Additionally, we could work on created a dedicated previewer that shows the data in a nicer way than raw XML.

The code we write will look something like this:

```
String ncml = netcdfFile.toNcml(file.getName());
```

Here's the output for an HDF5 file at `src/test/resources/hdf/hdf5/vlen_string_dset` (from the PR above):

```
<?xml version="1.0" encoding="UTF-8"?>
<netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2" location="file:vlen_string_dset">
  <variable name="DS1" shape="4" type="String" />
</netcdf>
```

Here's part of the output for a NetCDF file at `src/test/resources/netcdf/madis-raob.nc` (also from the PR above):

```
<?xml version="1.0" encoding="UTF-8"?>
<netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2" location="file:madis-raob.nc">
  <dimension name="recNum" length="1" isUnlimited="true" />
  <dimension name="manLevel" length="22" />
  <dimension name="sigTLevel" length="150" />
  <dimension name="sigWLevel" length="76" />
  <dimension name="mWndNum" length="4" />
  <dimension name="mTropNum" length="4" />
  <dimension name="staNameLen" length="50" />
  <dimension name="QCcheckNum" length="10" />
  <dimension name="QCcheckNameLen" length="60" />
  <dimension name="maxStaticIds" length="1000" />
  <dimension name="totalIdLen" length="50" />
  <dimension name="nInventoryBins" length="32" />
  <variable name="nStaticIds" shape="" type="int">
    <attribute name="_FillValue" type="int" value="0" />
  </variable>
  <variable name="staticIds" shape="maxStaticIds totalIdLen" type="char">
    <attribute name="_FillValue" value="" />
  </variable>
  <variable name="lastRecord" shape="maxStaticIds" type="int">
    <attribute name="_FillValue" type="int" value="-1" />
  </variable>
  <variable name="invTime" shape="recNum" type="int">
    <attribute name="_FillValue" type="int" value="0" />
  </variable>
...
```

Here's the full XML/NcML output: [madis-ncml.xml.txt](https://github.com/IQSS/dataverse/files/9972770/madis-ncml.xml.txt)

---

2.5 years ago @qqmyers made some suggestions for previewing XML files at https://github.com/IQSS/dataverse.harvard.edu/issues/70#issuecomment-631087257 . Here's his comment:

"FWIW: Something like https://www.jqueryscript.net/other/tree-xml-viewer-formatter.html adapted with the wiki instructions at https://github.com/GlobalDataverseCommunityConsortium/dataverse-previewers/wiki/How-to-create-a-previewer might be a quick win. (I didn't search too hard for an XML viewer - there could be better libraries out there to start from.)"

### PR DESCRIPTION
**Pull request by pdurbin**
**What this PR does / why we need it**:

From talking to researchers who use NetCDF and HDF5 files, we believe there is value in extracting metadata from those files to give some insight about their contents.

I plan to also make a pull request in the https://github.com/gdcc/dataverse-previewers repo to add a preview tool for the extracted file, which is in NcML format.

**Which issue(s) this PR closes**:

- Closes #9153
- Relates to #9242

**Special notes for your reviewer**:

Additional scope we could consider:

- an API endpoint to extract NcML from existing NetCDF and HDF5 files
- have the "requirements" option apply in more places than file level preview tools
- making the extraction feature modular

**Suggestions on how to test this**:

- try various NetCDF and HDF5 files and see if NcML aux files are created
- try the new "requirements" option for external tools
- use the proposed NcML previewer in this PR: https://github.com/gdcc/dataverse-previewers/pull/18

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

No but here's an example of two HDF5 files where one doesn't show the eyeball because it doesn't meet the requirement of having a aux file:

Screen Shot 2022-12-20 at 10 05 08 AM

**Is there a release notes update needed for this change?**:

Yes, included.

**Additional documentation**:

Included.

### CODE REVIEW

**Comment by pdurbin** [comment]
The failing test is this:

java.lang.AssertionError: expected:<201> but was:<500>
	at edu.harvard.iq.dataverse.api.HarvestingClientsIT.testHarvestingClientRun(HarvestingClientsIT.java:184)

I just clicked "build now" to see if that fixes it.

**Comment by mreekie** [comment]
Update. This was reviewed by Stephen, but is on hold pending some code tweaks from Phil.
Will go back to review by Stephen after that.

**Comment by mreekie** [comment]
update: Permissions are being discussed.
Stephen/Phil To keep this simple for now, superuser permissions will be required.

**Comment by pdurbin** [comment]
> an API endpoint to extract NcML from existing NetCDF and HDF5 files

I just added this in e2066c8.

I'm ready for more review so I'm unassigning myself.

File: src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java [hunk 1]
```diff
@@ -293,6 +305,7 @@ public List<DataFile> saveAndAddFilesToDataset(DatasetVersion version,
 					// Any necessary post-processing:
 					// performPostProcessingTasks(dataFile);
 				} else {
+                                    System.out.println("driver is not tmp");
```
Comment (sekmiller) [inline src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java:308]: Do we want to user logger here? warning or something else?

Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/ingest/IngestServiceBean.java]: Whoops! Fixed (removed) in 8cead71.

**Comment by pdurbin** [comment]
@sekmiller and talked a bit about the new "requirements" feature for external tools. I also removed an extraneous println he caught. Finally, I merged the latest from develop in.

**Review by sekmiller — APPROVED** [review]
Spoke with Phil to get a fuller understanding of "meetsRequirements" on the external tool. passing to qa

**Comment by mreekie** [comment]
size=10 for the next sprint

**Comment by pdurbin** [comment]
I'm not sure why I'm seeing this at https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-9239/8/consoleFull

TASK [dataverse : fire off installer] ******************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "/usr/bin/python3 /tmp/dvinstall/install.py -f --config_file=default.config --noninteractive > /tmp/dvinstall/install.out 2>&1", "delta": "0:02:23.416798", "end": "2023-01-11 23:03:08.721725", "msg": "non-zero return code", "rc": 1, "start": "2023-01-11 23:00:45.304927", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

**Comment by kcondon** [comment]
Generally works good. Found 2 issues:
1. Extracting metadata with api endpoint works on draft dataset but returns 404 on published one.
2. One test hdf file is processed and a NcML aux file is produced but not displaying eye icon for auxfiles previewer tool.

Will retest 1 in case I made a mistake at the end of the day. 2. may be an edge case since all my other test files worked as expected but wanted to mention it.
